src/vcpkg/base/system.process.cpp

```diff
@@ -476,18 +476,21 @@ namespace vcpkg
         }
 
         std::atomic<size_t> work_item{0};
-        const auto num_threads =
-            static_cast<size_t>(std::max(1, std::min(get_concurrency(), static_cast<int>(cmd_lines.size()))));
+        std::mutex mtx;
+        const auto num_threads = std::max(static_cast<size_t>(1), std::min(static_cast<size_t>(get_concurrency()), cmd_lines.size()));
 
         auto work = [&]() {
             std::size_t item;
             while (item = work_item.fetch_add(1), item < cmd_lines.size())
             {
-                res[item] = cmd_execute_and_capture_output(cmd_lines[item], wd, env);
```

Comment:
Crash because of race condition

Reply:
I'm really wondering if this is it, though. Maybe I'm wrong, but I've always been under the impression that updating _distinct keys_ from an `std::vector` from multiple threads should be safe, provided that you're not resizing the `std::vector`.

Once it's allocated, it's just a wrapper around a chunk of memory, as long as writes touch different memory addresses I'd imagine that should be safe to happen simultaneously.

Reply:
> Once it's allocated, it's just a wrapper around a chunk of memory, as long as writes touch different memory addresses I'd imagine that should be safe to happen simultaneously.

Technically, I would agree with you. However, I don't think that the standard specifies that there shall be no resizing going on between vector allocation and between insertions at defined indexes.

The other possibility is that multiple threads try to write to the same index but this is permitted by the atomic add.

I think this is a use case for parallel `for_each` but unfortunately, Clang doesn't support this.

Reply:
Additionally, I don't think that the lock will hurt performance that much because `operator[]` should be fast in comparison to executing the command.

Reply:
> I'm really wondering if this is it, though. Maybe I'm wrong, but I've always been under the impression that updating _distinct keys_ from an `std::vector` from multiple threads should be safe, provided that you're not resizing the `std::vector`.
> 
> Once it's allocated, it's just a wrapper around a chunk of memory, as long as writes touch different memory addresses I'd imagine that should be safe to happen simultaneously.

@omartijn I think you are actually right there. I did some tests and it seems like concurrent writes to different indexes of the vector aren't a problem at all here and I can safely remove the lock. I did some research on the error code `EXC_BAD_ACCESS` and `KERN_INVALID_ADDRESS` and found issues that point to bugs in Xcode. Since the crash occurs in `popen` I think that this is not an issue with vcpkg but related to Xcode. The correct thing would be to compile vcpkg with a newer version of Xcode and hope that the bug is fixed. I will also add unit tests to make sure we don't mess up concurrency here.

Reply:
If it is indeed an xcode bug, that would explain why I was unable to reproduce it with a local debug build. I'm interested which issues you found that it's an issue in xcode. Could you share links?

The fact that it crashes inside `popen` itself is not an indication that it's a bug in xcode. In fact, `popen` is part of the system library and is not updated by an xcode update. Usually a `popen` crash means that one of the arguments is garbage (i.e. `nullptr` or already freed, not correctly null-terminated).